### PR TITLE
nodesConfig pkg added (nodes.config object)

### PR DESCRIPTION
**Pull request by elenagerman**
nodes.config object pkg

### CODE REVIEW

**Comment by kononovn** [comment]
Please add unit test coverage. 

Link for more info: https://github.com/openshift-kni/eco-goinfra#note-every-new-package-requires-a-coverage-of-all-its-public-functions-with-unit-tests-unit-tests-are-located-in-the-same-package-as-the-resource-in-a-file-with-the-name-resource_testgo-examples-can-be-found-in-configmap_testgo-and-metallb_testgo

**Comment by elenagerman** [comment]
> Please add unit test coverage.
> 
> Link for more info: https://github.com/openshift-kni/eco-goinfra#note-every-new-package-requires-a-coverage-of-all-its-public-functions-with-unit-tests-unit-tests-are-located-in-the-same-package-as-the-resource-in-a-file-with-the-name-resource_testgo-examples-can-be-found-in-configmap_testgo-and-metallb_testgo

sure, but few questions before:
1. why you didn't ask about it for the apiservers and for the imageregistry pkg?
2. nodes.config with the name cluster is also created during cluster deployment and can be only edited (like changing cgroup value) - do we still need a unit test?

**Comment by elenagerman** [comment]
@kononovn all I can do is just to create an empty _test file

File: pkg/nodesconfig/nodesconfig.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	Object *configV1.Node
+	// api client to interact with the cluster.
+	apiClient goclient.Client
+	// Used in functions that define or mutate clusterOperator definition. errorMsg is processed before the
+	// ClusterOperator object is created.
+	errorMsg string
+}
+
+// NewBuilder method creates new instance of builder.
+func NewBuilder(apiClient *clients.Settings, name string, cgroupMode configV1.CgroupMode) *Builder {
```
[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go:31]: sure, np!

File: pkg/nodesconfig/nodesconfig.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		return nil, fmt.Errorf("nodesConfig Config 'apiClient' cannot be empty")
+	}
+
+	if nodesConfigObjName == "" {
+		glog.V(100).Infof("The name of the nodesConfig is empty")
+
+		return nil, fmt.Errorf("nodesConfig 'nodesConfigObjName' cannot be empty")
+	}
+
+	glog.V(100).Infof(
```
Comment (kononovn) [inline pkg/nodesconfig/nodesconfig.go:44]: Please move this logs to the first like of this function

File: pkg/nodesconfig/nodesconfig.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	err := builder.apiClient.Update(context.TODO(), builder.Definition)
+	if err == nil {
+		builder.Object = builder.Definition
+	}
+
+	return builder, err
+}
+
+// GetNodesConfigIoGVR returns nodesConfig's GroupVersionResource which could be used for Clean function.
+func GetNodesConfigIoGVR() schema.GroupVersionResource {
```
Comment (kononovn) [inline pkg/nodesconfig/nodesconfig.go:122]: Please move this standalone function down and put it under > WithCGroupMode.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please address few comments and I'll add lgtm, thanks.

File: pkg/nodesconfig/nodesconfig.go (hunk 2 above)
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go:44]: not sure what you ask about, but pls check

File: pkg/nodesconfig/nodesconfig.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+// WithCGroupMode sets the nodesConfig operator's cgroupMode.
+func (builder *Builder) WithCGroupMode(expectedCGroupMode configV1.CgroupMode) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if expectedCGroupMode == configV1.CgroupModeEmpty {
+		builder.errorMsg = "the cGroup mode value can not be empty"
```
[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go:150]: is it ok?

File: pkg/nodesconfig/nodesconfig.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+		return builder
+	}
+
+	glog.V(100).Infof(
```
[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go]: is it ok?

File: pkg/nodesconfig/nodesconfig_test.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+			assert.NotNil(t, err)
+			assert.Equal(t, testCase.expectedError, err)
+		} else {
+			assert.Nil(t, err)
+			assert.Equal(t, testNodeConfig.Name, builderResult.Object.Name)
+		}
+	}
+}
+
+func TestNodesConfigNewBuilder(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig_test.go:102]: sure, thanks, I missed it

File: pkg/nodesconfig/nodesconfig.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kononovn) [reply pkg/nodesconfig/nodesconfig.go:146]: I don't see glog.V here. Can you add log ?

[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go:146]: aa, ok, I see what you mean

**Review by kononovn — APPROVED** [review]
/lgtm

**Comment by elenagerman** [comment]
@trewest @achuzhoy @cdvultur PTAL

**Review by sebrandon1 — APPROVED** [review]

File: pkg/nodesconfig/nodesconfig.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		return "", fmt.Errorf("imageRegistry object doesn't exist")
+	}
+
+	return builder.Object.Spec.CgroupMode, nil
+}
+
+// WithCGroupMode sets the nodesConfig operator's cgroupMode.
+func (builder *Builder) WithCGroupMode(expectedCGroupMode configV1.CgroupMode) *Builder {
+	glog.V(100).Infof("Setting nodesConfig %s with ManagementState: %v",
+		builder.Definition.Name, expectedCGroupMode)
```
Comment (trewest) [inline pkg/nodesconfig/nodesconfig.go:139]: Please add validate block before this

**Review by trewest — CHANGES_REQUESTED** [review]
Looks good, just one comment

File: pkg/nodesconfig/nodesconfig.go (hunk 7 above)
Comment (elenagerman) [reply pkg/nodesconfig/nodesconfig.go:139]: we have it just after the glog. I'll switch them...

Comment (trewest) [reply pkg/nodesconfig/nodesconfig.go]: Yeah but the validate method protects us from nil pointer exceptions

**Review by trewest — APPROVED** [review]
/lgtm

**Pull request merged**